### DISCUSSION TITLE
Zod enum from keys of an object

### DISCUSSION THREAD

**Discussion by bennik88** [colinhacks/zod#2575] [Q&A]
### Problem
I don't want to type my constants in this case "EN", "JA" multiple times over my codebase so I want to source the types and enums from `languageAbbreviations` but I get a typescript or zod error depending on what I try. This seems to be a common case. How would I solve this.

### Example:
```js
type LanguageAbbreviationKeys = keyof typeof languageAbbreviations;
type LanguageAbbreviationValues = (typeof languageAbbreviations)[keyof typeof languageAbbreviations];

const languageAbbreviations = {
  EN: "English",
  JA: "Japanese",
} as const;

const languageKeys = Object.keys(languageAbbreviations) as LanguageAbbreviationKeys[];

const validator = z.object({
  text: z.string(),
 // Here i want to infer the enum and types
  target_language: z.nativeEnum(languageKeys)
});
```
### What I've tried:

I tried multiple solutions one of them was a custom function. But as it's not typesafe it will fail on the return statement when linting.
```js
function zodEnumFromObjKeys<K extends string> (obj: Record<K, any>): z.ZodEnum<[K, ...K[]]> {
  const [firstKey, ...otherKeys] = Object.keys(obj) as K[]
  return z.enum([firstKey, ...otherKeys])
}
```

**Comment by feychenie** [comment]
Proposed an answer to your original comment/question here: 
https://github.com/colinhacks/zod/discussions/839#discussioncomment-6488540

**Comment by danny-does-stuff** [comment]
I got the types I was hoping for using this function

```
	function getZodEnumFromObjectKeys<K extends string>(
		input: Record<K, unknown>,
	): z.ZodEnum<{[k in K]: k}> {
		return z.enum(Object.keys(input) as K[])
	}
```
